src/main/java/vg/civcraft/mc/citadel/activity/ActivityMap.java

```diff
@@ -0,0 +1,209 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.World;
+import vg.civcraft.mc.citadel.Citadel;
+import vg.civcraft.mc.civmodcore.CivModCorePlugin;
+import vg.civcraft.mc.civmodcore.dao.ManagedDatasource;
+import vg.civcraft.mc.civmodcore.locations.global.WorldIDManager;
+import vg.civcraft.mc.namelayer.group.Group;
+
+public class ActivityMap {
+
+	private static final String CREATE_TABLE = "CREATE TABLE IF NOT EXISTS ctdl_activity_map ("
+			+ "group_id INT, "
+			+ "world SMALLINT, "
+			+ "x INT, "
+			+ "z INT, "
+			+ "resolution INT, "
+			+ "activity TIMESTAMP, "
+			+ "PRIMARY KEY(group_id, world, x, z, resolution))";
+	private static final String GET_ACTIVITY = "SELECT x, z, activity FROM ctdl_activity_map WHERE group_id = ? AND world = ? AND resolution = ?";
+	private static final String UPDATE_ACTIVITY = "INSERT INTO ctdl_activity_map (group_id, world, x, z, resolution, activity) VALUES (?, ?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE activity = ?";
+
+	private final WorldIDManager worldIdManager;
+	// Must be array list so we can resize it
+	private final Map<Short, LoadingCache<Integer, Table<Integer, Integer, Instant>>> activityTimes;
+	private final Queue<Update> updates = new LinkedBlockingQueue<>();
+
+	private final List<String> worlds;
+	private final int resolution;
+	private final int radius;
+	private final Instant defaultActivity;
+
+	private final ManagedDatasource source;
+
+	public ActivityMap(ManagedDatasource source) {
+		this.source = source;
+		worldIdManager = CivModCorePlugin.getInstance().getWorldIdManager();
+		activityTimes = new ConcurrentHashMap<>();
+
+		resolution = Citadel.getInstance().getConfigManager().getActivityMapResolution();
+		radius = Citadel.getInstance().getConfigManager().getActivityMapRadius();
+		worlds = Citadel.getInstance().getConfigManager().getActivityWorlds();
+		defaultActivity = Instant
+				.ofEpochSecond(Citadel.getInstance().getConfigManager().getActivityDefault());
+
+		Bukkit.getScheduler().runTaskTimerAsynchronously(Citadel.getInstance(), () -> {
+			try (Connection connection = source.getConnection()) {
+				PreparedStatement statement = connection.prepareStatement(UPDATE_ACTIVITY);
+
+				Update update;
+				while ((update = updates.poll()) != null) {
+					statement.setInt(1, update.group);
+					statement.setShort(2, update.world);
+					statement.setInt(3, update.x);
+					statement.setInt(4, update.z);
+					statement.setInt(5, resolution);
+					statement.setTimestamp(6, Timestamp.from(update.timestamp));
+					statement.setTimestamp(7, Timestamp.from(update.timestamp));
+					statement.addBatch();
+				}
+
+				statement.executeBatch();
+			} catch (SQLException ex) {
+				ex.printStackTrace();
+			}
+			//}, 20 * 60 * 4, 20 * 60 * 4);
+		}, 20 * 30, 20 * 30);
+
+		try (Connection connection = source.getConnection()) {
+			connection.createStatement().executeUpdate(CREATE_TABLE);
```

Comment:
Should be a migration

Reply:
explain?

Reply:
https://github.com/CivClassic/Citadel/blob/master/src/main/java/vg/civcraft/mc/citadel/model/CitadelDAO.java#L108

We have a system in place specifically for adding/modifying tables. This allow better tracking of when things break and modifying tables later on.